agent/invocation/state_machine.rb

```diff
@@ -0,0 +1,105 @@
+require "forwardable"
+require "micromachine"
+require_relative "../agent"
+
+module FastlaneCI::Agent
+  ##
+  # StateMachine defines the state and transition events of an Invocation
+  #
+  # This module is implemented to be `prepended` into a class.
+  #
+  # Methods for each transition are defined, and will only be called in the containing class
+  # if the transition is valid.
+  #
+  # On every transition `send_status` will be attempted to be called.
+  module StateMachine
+    include Logging
+    extend Forwardable
+
+    # forward methods to the state machine
+    def_delegators :state_machine, :state, :events, :states
+
+    def state_machine
+      @state_machine ||= MicroMachine.new("pending").tap do |fsm|
+        fsm.when(:run,     "pending"   => "running")
```

Comment:
👍 good idea

Reply:
This doesn't look like it's going to be very easy, since the enum values are integers and the state machine accepts strings as fields. It may be possible to use some lookup methods to get the values you need, but I think it reduces readability substantially. I think it's fine to keep the hardcoded strings as long as we have a test to ensure these strings match the enums:

https://github.com/fastlane/ci/pull/987/files#diff-cfa28a249b00816d6bd72b264ead3a71R17